### ISSUE TITLE
"Marker points into wrong buffer" with helm-mark-ring

### ISSUE THREAD

**Issue by Ambrevar** [emacs-evil/evil#845]
# Issue type

- Bug report

# Environment

Emacs version: GNU Emacs 25.2.1 (x86_64-unknown-linux-gnu, GTK+ Version 3.22.10) of 2017-04-22
Operating System: Arch Linux
Evil version: 1.2.12
Evil installation type: MELPA
Graphical/Terminal: Graphical
Terminal multiplexer: N/A

# Reproduction steps

Install latest helm.

Start Emacs, require helm and evil.

Use some vim movements (e.g. `{` and `}`) to populate the mark ring.

Call `helm-mark-ring` and make sure you have at least 2 entries.

Enable _helm-follow-mode_ with `C-c C-f`.

Move around the helm entries up and down until you get the following error

	Marker points into wrong buffer

**Comment by wasamasa** [comment]
I cannot reproduce this.  Are you sure this is Evil-specific?  The mark is not a Vim invention, so I'd expect the issue to show up when using vanilla Emacs commands, such as C-SPC.

**Comment by Ambrevar** [comment]
You might have to browse the entries a dozen of times.
If it does not happen to you, I'm not sure what I could be missing.

When no evil movement has been done and only Emacs-marks are in the ring (e.g. with C-SPC), the issue does not occur.

Here is the trace:
```
Debugger entered--Lisp error: (error "Marker points into wrong buffer" #<marker in no buffer>)
  helm-goto-char(#<marker in no buffer>)
  helm-mark-ring-default-action(#<marker in no buffer>)
  helm-execute-selection-action-1(#<marker in no buffer> (("Goto line" . helm-mark-ring-default-action)) t)
  helm-execute-persistent-action()
  #[0 "\205\301 \207" [helm-alive-p helm-execute-persistent-action] 1]()
  apply(#[0 "\205\301 \207" [helm-alive-p helm-execute-persistent-action] 1] nil)
  timer-event-handler([t 0 0 10000 nil #[0 "\205\301 \207" [helm-alive-p helm-execute-persistent-action] 1] nil idle 0])
  read-from-minibuffer("pattern: " <TRUNCATED>
  helm-read-pattern-maybe(nil nil nil noresume nil nil nil)
  helm-internal((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil)
  apply(helm-internal ((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil))
  helm((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil)
  apply(helm ((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil))
  helm(:sources (helm-source-mark-ring helm-source-global-mark-ring) :resume noresume :buffer "*helm mark ring*")
  helm-all-mark-rings()
  (if rectangle-mark-mode (rectangle-exchange-point-and-mark) (helm-all-mark-rings))
  helm-mark-or-exchange-rect()
  funcall-interactively(helm-mark-or-exchange-rect)
  call-interactively(helm-mark-or-exchange-rect nil nil)
  command-execute(helm-mark-or-exchange-rect)
```

And the mark-ring:
```
(#<marker at 190 in bar.c> #<marker at 143 in bar.c> #<marker at 190 in bar.c> #<marker at 143 in bar.c> #<marker at 190 in bar.c> #<marker at 23 in bar.c> #<marker at 190 in bar.c> #<marker at 143 in bar.c> #<marker at 190 in bar.c> #<marker at 143 in bar.c> #<marker at 190 in bar.c> #<marker at 23 in bar.c> #<marker at 190 in bar.c> #<marker at 143 in bar.c> #<marker at 245 in bar.c> #<marker at 245 in bar.c>)
```

**Comment by wasamasa** [comment]
Provide a minimal repro, then we'll talk.

**Comment by Ambrevar** [comment]
- Start `emacs -Q`.
- Install latest Evil and Helm.
- Run the following:

		(require 'helm-config)
		(require 'evil)

	No need to turn on Evil mode.

- Make two different marks with `C-SPC`.

- `M-x helm-mark-ring`. (Make sure you have at least 2 entries.)

- Enable `helm-follow-mode` with `C-c C-f`.

- Move around the Helm entries up and down (`C-p`/`C-n`) until you get the error.

**Comment by Ambrevar** [comment]
Trace (the other one is not exact):

```
(error "Marker points into wrong buffer" #<marker in no buffer>)
  helm-goto-char(#<marker in no buffer>)
  helm-mark-ring-default-action(#<marker in no buffer>)
  helm-execute-selection-action-1(#<marker in no buffer> (("Goto line" . helm-mark-ring-default-action)) t)
  helm-execute-persistent-action()
  #[0 "�\205�^@\301 \207" [helm-alive-p helm-execute-persistent-action] 1]()
  apply(#[0 "�\205�^@\301 \207" [helm-alive-p helm-execute-persistent-action] 1] nil)
  timer-event-handler([t 0 0 10000 nil #[0 "�\205�^@\301 \207" [helm-alive-p helm-execute-persistent-action] 1] nil idle 0])
  read-from-minibuffer("pattern: " nil (keymap (f13 lambda nil (interactive) (helm-select-nth-action 12)) (f12 lambda nil (interactive) (helm-select-nth-action 11)) (f11 lambda nil (interactive) (helm-select-nth-action 10)) (f10 lambda nil (interactive) (helm-select-nth-action 9)) (f9 lambda nil (interactive) (helm-select-nth-action 8)) (f8 lambda nil (interactive) (helm-select-nth-action 7)) (f7 lambda nil (interactive) (helm-select-nth-action 6)) (f6 lambda nil (interactive) (helm-select-nth-action 5)) (f5 lambda nil (interactive) (helm-select-nth-action 4)) (f4 lambda nil (interactive) (helm-select-nth-action 3)) (f3 lambda nil (interactive) (helm-select-nth-action 2)) (f2 lambda nil (interactive) (helm-select-nth-action 1)) (menu-bar keymap (help-menu keymap (describe keymap (describe-mode . helm-help)))) (help keymap (109 . helm-help)) (23 . #[0 "\306\307\310\311\312\305!\313\"\314$\216\315\302!\210\316\304!\317\211\211����\301=\203'^@\315\302!\210\320\202]^@\321�\303\"\211\262�\203:^@\315�A!\210\320\202]^@\312��!\262�\322\300�\"\262�\323�!\203S^@\315�!\210\202\\^@\324\325�\"��\244��\317\266\204\205e^@\202�^@)\207" [#0 23 helm-yank-text-at-point ((31 . helm-undo-yank-text-at-point)) nil nil make-byte-code 0 "\300\205�^@\300 \207" vconcat vector [] 1 call-interactively read-key nil t assoc lookup-key commandp mapcar identity last-command-event unread-command-events] 7 nil nil]) (f1 lambda nil (interactive) (helm-select-nth-action 0)) (8 keymap (109 . helm-help) (104 . undefined) (8 . undefined) (99 . helm-customize-group) (4 . helm-enable-or-switch-to-debug)) (20 . helm-toggle-resplit-and-swap-windows) (C-tab . undefined) (67108897 . helm-toggle-suspend-update) (3 keymap (57 . helm-execute-selection-action-at-nth-+9) (56 . helm-execute-selection-action-at-nth-+8) (55 . helm-execute-selection-action-at-nth-+7) (54 . helm-execute-selection-action-at-nth-+6) (53 . helm-execute-selection-action-at-nth-+5) (52 . helm-execute-selection-action-at-nth-+4) (51 . helm-execute-selection-action-at-nth-+3) (50 . helm-execute-selection-action-at-nth-+2) (49 . helm-execute-selection-action-at-nth-+1) (63 . helm-help) (110 . #[0 "\306\307\310\311\312\305!\313\"\314$\216\315\302!\210\316\304!\317\211\211����\301=\203'^@\315\302!\210\320\202]^@\321�\303\"\211\262�\203:^@\315�A!\210\320\202]^@\312��!\262�\322\300�\"\262�\323�!\203S^@\315�!\210\202\\^@\324\325�\"��\244��\317\266\204\205e^@\202�^@)\207" [#0 110 helm-run-cycle-resume nil nil nil make-byte-code 0 "\300\205�^@\300 \207" vconcat vector [] 1 call-interactively read-key nil t assoc lookup-key commandp mapcar identity last-command-event unread-command-events] 7 nil nil]) (62 . helm-toggle-truncate-line) (21 . helm-refresh) (6 . helm-follow-mode) (9 . helm-copy-to-buffer) (11 . helm-kill-selection-and-quit) (25 . helm-yank-selection) (45 . helm-swap-windows)) (67108987 . helm-enlarge-window) (67108989 . helm-narrow-window) (19 . undefined) (24 keymap (57 . helm-execute-selection-action-at-nth-+9) (56 . helm-execute-selection-action-at-nth-+8) (55 . helm-execute-selection-action-at-nth-+7) (54 . helm-execute-selection-action-at-nth-+6) (53 . helm-execute-selection-action-at-nth-+5) (52 . helm-execute-selection-action-at-nth-+4) (51 . helm-execute-selection-action-at-nth-+3) (50 . helm-execute-selection-action-at-nth-+2) (49 . helm-execute-selection-action-at-nth-+1) (2 . helm-resume-list-buffers-after-quit) (98 . helm-resume-previous-session-after-quit) (6 . helm-quit-and-find-file)) (11 . helm-delete-minibuffer-contents) (67108896 . helm-toggle-visible-mark) (0 . helm-toggle-visible-mark) (C-M-up . helm-scroll-other-window-down) (C-M-down . helm-scroll-other-window) (M-prior . helm-scroll-other-window-down) (M-next . helm-scroll-other-window) (12 . helm-recenter-top-bottom-other-window) (15 . helm-next-source) (10 . helm-execute-persistent-action) (26 . helm-execute-persistent-action) (9 . helm-select-action) (13 . helm-maybe-exit-minibuffer) (left . helm-previous-source) (right . helm-next-source) (7 . helm-keyboard-quit) (22 . helm-next-page) (27 keymap (110 . next-history-element) (112 . previous-history-element) (115 . undefined) (5 . helm-display-all-sources) (1 . helm-show-all-in-this-source-only) (85 . helm-unmark-all) (97 . helm-mark-all) (109 . helm-toggle-all-marks) (41 . helm-next-visible-mark) (40 . helm-prev-visible-mark) (91) (32 . helm-toggle-visible-mark) (33554454 . helm-scroll-other-window-down) (25 . helm-scroll-other-window-down) (22 . helm-scroll-other-window) (12 . helm-reposition-window-other-window) (111 . helm-previous-source) (62 . helm-end-of-buffer) (60 . helm-beginning-of-buffer) (118 . helm-previous-page)) (next . helm-next-page) (prior . helm-previous-page) (C-up . helm-follow-action-backward) (C-down . helm-follow-action-forward) (16 . helm-previous-line) (14 . helm-next-line) ...) nil nil nil t)
  helm-read-pattern-maybe(nil nil nil noresume nil nil nil)
  helm-internal((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil)
  apply(helm-internal ((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil))
  helm((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil)
  apply(helm ((helm-source-mark-ring helm-source-global-mark-ring) nil nil noresume nil "*helm mark ring*" nil nil nil))
  helm(:sources (helm-source-mark-ring helm-source-global-mark-ring) :resume noresume :buffer "*helm mark ring*")
  helm-all-mark-rings()
  funcall-interactively(helm-all-mark-rings)
  call-interactively(helm-all-mark-rings record nil)
  command-execute(helm-all-mark-rings record)
  execute-extended-command(nil "helm-all-mark-rings" nil)
  funcall-interactively(execute-extended-command nil "helm-all-mark-rings" nil)
  call-interactively(execute-extended-command nil nil)
  command-execute(execute-extended-command)
```

**Comment by wasamasa** [comment]
Hm, this I can repro. I've started digging a bit more into this and found that Helm extracts the associated buffer from markers without checking whether one can actually switch to it, so considering that its code is extremely convoluted and the issue only occasionally happens when a timer fires, I'm not willing to fix their sloppy coding.

**Comment by Ambrevar** [comment]
Thanks for looking into this.

I could help fixing Helm, but your help is also invaluable.  Can you provide some more details?

- Which timer?

- Checking if a buffer can be switched to sounds reasonable, but why would it be an issue in `helm-mark-ring` which acts upon the current buffer only?

- What do you mean with "occasionally"?  In my case it happens extremely often to the point `helm-mark-ring` is unusable.

**Comment by wasamasa** [comment]
- See the backtrace.
- It doesn't.  It retrieves markers, creates markers on its own and follows whatever buffer the marker to jump to (thanks to its follow mode) points to, for some reason that marker doesn't have one
- The bug isn't triggered by precisely one action, but by repeating actions until they interfere with a timer run at who knows what interval.  This behavior is characteristic of races and those are a pain to debug.  I therefore leave this to the author of the package.

**Comment by Ambrevar** [comment]
Thank you, I think that will be very helpful.

About your last point: yes, it looks very much like a race condition, but the funny thing is that it is only triggered when evil is `require`d.  Does Evil do anything like setting up timers when the _file_ is loaded?  From `timer-list` and `timer-idle-list`, I cannot see such things.  Can you think of any side-effect?

**Comment by wasamasa** [comment]
The only thing I can think of is the recently rewritten `evil-search` code cleaning up its markers by unsetting their buffer.  This is one of the few best practices in elisp to follow as they can otherwise not be garbage-collected.

**Comment by Ambrevar** [comment]
Very good, I'll look into it.

**Comment by Ambrevar** [comment]
I've inspected the state of the `mark-ring` (without Helm), and I've noticed something that maybe wrong: it seems that whenever Evil jumps (with `C-i`/`C-o`), it puts a mark on the current point before the jump, _even if_ it is already in the mark ring.  Which means a succession of jumps will keep populating the ring.

Recipe on a new file:
- Insert

		foo

		bar

		baz

	The point is now after the `z` of `baz`, that is pos 14.

- Press `{`.  The `(mark)` should now be at 14 and the mark-ring empty.

- Press `{`: the point is 5, the mark is 10, the mark-ring has `14`.

- Press `C-o`: the point is 10, the mark is 5, the mark-ring is `10 14`.
- Press `C-i`: the point is 5, the mark is 10, the mark-ring is `5 10 14`.
- Press `C-o`: the point is 10, the mark is 5, the mark-ring is `10 5 10 14`.
- Press `C-i`: the point is 5, the mark is 10, the mark-ring is `5 10 5 10 14`.
- Etc.
